Contribute svcat

carolynvs opened the pull request:
This moves svcat out of [Azure/service-catalog-cli](https://github.com/Azure/service-catalog-cli) and into this repo. 

**Package Structure**
* cmd/svcat - The cli shell.
* cmd/svcat/plugin - Handles plugin related concerns, like mapping environment variables passed from kubectl to the command flags.
* pkg/svcat - The svcat app.

**Commands**
* `install plugin` - Copies the binary to the plugins directory, and generates a plugin manifest.
* `get` and `describe` for broker, class, plan, instance and binding
* `sync broker`
* `provision`, `deprovision`, `bind`, `unbind`

**Dev Tasks**
* `go build ./cmd/svcat` compiles the binary.
* `go test ./cmd/svcat/...` runs the unit tests. I've been working on integration tests but those aren't ready to contribute yet.

**TODO/Missing**
I'm just calling out things that would probably need to happen, if this is accepted.

* Add build and test calls to the makefile
* Change how versioning is done for the binary, right now it relies on flags during compilation: `-X main.commit=$(COMMIT) -X main.version=$(VERSION)` where commit and version came from git.
* Publish the binaries to github release page.
* Update installation instructions to point to the new release location.

MHBauer commented:
Do you want to try to keep all the history of development? I wouldn't want people to miss out on that.

carolynvs commented:
I don't have strong feelings on the git history. Honestly, I'm not even sure how hard/possible it is to move code across repos with different structures and keep the history... 😀

arschles commented:
@carolynvs looks like the CI failure is legit: https://travis-ci.org/kubernetes-incubator/service-catalog/builds/327880013#L574

carolynvs commented:
Yes, how emabarrasing! 😉 I found the makefile and will use that to get the code up to par.

jberkhahn commented:
My only concern is I would like some more robust unit tests. What we have so far seems pretty spartan.

MHBauer commented:
Looks like it's @carolynvs @arschles & @sozercan who have all the history. If you're comfortable losing the history, that's fine with me. It's possible and not too hard to keep it if you're interested.

pmorie commented:
Are there license issues that arise from retaining the git history?

On Thu, Jan 11, 2018 at 7:27 PM, Morgan Bauer <notifications@github.com>
wrote:

> Looks like it's @carolynvs <https://github.com/carolynvs> @arschles
> <https://github.com/arschles> & @sozercan <https://github.com/sozercan>
> who have all the history. If you're comfortable losing the history, that's
> fine with me. It's possible and not too hard to keep it if you're
> interested.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-incubator/service-catalog/pull/1664#issuecomment-357106045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWXmAa07BvvCjE2S4aYPE3X8OYXDTAbks5tJqbdgaJpZM4RbeE2>
> .
>

n3wscott commented on cmd/svcat/README.md:
````diff
@@ -0,0 +1,211 @@
+
+## Bash
+```
+curl -sLO https://servicecatalogcli.blob.core.windows.net/cli/latest/$(uname -s)/$(uname -m)/svcat
````
Will the built versions continue to be built and hosted by this URI after the tool have been moved into service-catalog?

pmorie replied:
This is a very good point.

carolynvs replied:
I mentioned in the OP that these would need to be changed once we start building and publishing binaries during the build.

carolynvs commented on cmd/svcat/class/get_cmd.go:
```diff
@@ -0,0 +1,98 @@
+		Short:   "List classes, optionally filtered by name",
+		Example: `
+  svcat get classes
+  svcat get class azure-mysqldb
```
Working on that now! 👍 Switching everything over to the user provided service broker.

carolynvs commented:
> My only concern is I would like some more robust unit tests. What we have so far seems pretty spartan.

I am filling out the unit tests today. I had a branch where I was working on integration tests for svcat, but I left it out of this since a) it hadn't landed in svcat's master branch yet and b) it probably needs to change anyway to work with service catalog's integration test suite setup.

I can do a follow-on issue/PR to hook up the integration tests.

pmorie commented:
> I can do a follow-on issue/PR to hook up the integration tests.

Honestly, I would be okay with the unit tests in a follow-up, assuming that everyone else is. I'll leave that to you all to sort out. In general I tend to have a bias toward 'merge now, fix things in follow-up' when it comes to code donations, but that's up to you all.

pmorie commented on cmd/svcat/class/describe_cmd.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+// NewDescribeCmd builds a "svcat describe class" command
+func NewDescribeCmd(cxt *command.Context) *cobra.Command {
```
This is probably another one that we should rename away (in a follow-up, not this PR) from the kubectl verb to something else, maybe inspect?

carolynvs commented:
Let's do the CYA wording in the helptext on a follow-on PR. I'd like more time to come up with the proper wording for that.

carolynvs commented:
I've addressed the following items in this PR:
* Added unit tests for remaining idempotent commands
* Removed azure from the golden files and help text examples

The remaining items will be addressed in follow-on PRs:
* Publish binaries
  * Set the version
  * Publish binaries to github releases during CI
  * Update install doc with new release URLs
* Documentation
  * Move cli readme into /docs
  * Copy binary install instructions into install.md
  * Update walkthrough to use binary
* Clarify support status of the binary

carolynvs commented on cmd/svcat/class/get_cmd.go:
```diff
@@ -0,0 +1,98 @@
+	"github.com/spf13/cobra"
+)
+
+type getCmd struct {
```
It was a deliberate choice to keep as closely as possible to kubectl's grammar, which drove the choice not only of the verbs get and describe, but also using the same "binary verb noun" pattern.

It makes for a much easier learning curve when people's default instincts honed from using kubectl _just work_ with svcat. The output of svcat's get and describe were patterned after kubectl's output (though I didn't take it all the way due to time constraints).

I want to line-up svcat so that eventually when get and describe work well natively in kubectl, we can delegate those commands back to kubectl. Ideally resulting in as as few changes as possible for the end user, so they can seamlessly switch from `svcat get broker foo` to `kubectl get broker foo`.

pmorie commented:
@carolynvs looks like this needs rebase now

carolynvs commented:
Yup! I am working this morning on rebasing my PRs after the 1.9 bump went in.

carolynvs commented:
Rebased and the checks are passing.

carolynvs commented:
I found an existing bug report for kubectl not supporting boolean flags (such as `kubectl plugin svcat --version`): https://github.com/kubernetes/kubernetes/issues/53640.

kibbles-n-bytes commented:
I'm going to go ahead and pull the trigger on this; looks like it's in good enough shape to pull into the repo, and we can mess with it all we want in follow-ups.
